File: pkg/dev_compiler/test/codegen_expected/corelib/expression_test.js
dart_library.library('corelib/expression_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__expression_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const expression_test = Object.create(null);
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  expression_test.ExpressionTest = class ExpressionTest extends core.Object {
    new() {
      this.foo = null;
    }
    static testMain() {
      let test = new expression_test.ExpressionTest();
      test.testBinary();
      test.testUnary();
      test.testShifts();
      test.testBitwise();
      test.testIncrement();
      test.testMangling();
    }
    testBinary() {
      let x = 4, y = 2;
      expect$.Expect.equals(6, x + y);
      expect$.Expect.equals(2, x - y);
      expect$.Expect.equals(8, x * y);
      expect$.Expect.equals(2, x / y);
      expect$.Expect.equals(0, x[dartx['%']](y));
    }
    testUnary() {
      let x = 4, y = 2, z = -5;
      let t = true, f = false;
      expect$.Expect.equals(-4, -x);
      expect$.Expect.equals(4, ~z >>> 0);
      expect$.Expect.equals(f, !t);
    }
    testShifts() {
      let x = 4, y = 2;
      expect$.Expect.equals(y, x[dartx['>>']](1));
      expect$.Expect.equals(x, y << 1 >>> 0);
    }
    testBitwise() {
      let x = 4, y = 2;
      expect$.Expect.equals(6, (x | y) >>> 0);
      expect$.Expect.equals(0, (x & y) >>> 0);
      expect$.Expect.equals(6, (x ^ y) >>> 0);
    }
    get(index) {
      return this.foo;
    }
    set(index, value) {
      this.foo = value;
      return value;
    }
    testIncrement() {
      let x = 4, a = x++;
      expect$.Expect.equals(4, a);
      expect$.Expect.equals(5, x);
      expect$.Expect.equals(6, ++x);
      expect$.Expect.equals(6, x++);
      expect$.Expect.equals(7, x);
      expect$.Expect.equals(6, --x);
      expect$.Expect.equals(6, x--);
      expect$.Expect.equals(5, x);
      this.foo = 0;
      expect$.Expect.equals(0, (() => {
        let x = this.foo;
        this.foo = dart.notNull(x) + 1;
        return x;
      })());
      expect$.Expect.equals(1, this.foo);
      expect$.Expect.equals(2, (this.foo = dart.notNull(this.foo) + 1));
      expect$.Expect.equals(2, this.foo);
      expect$.Expect.equals(2, (() => {
        let x = this.foo;
        this.foo = dart.notNull(x) - 1;
        return x;
      })());
      expect$.Expect.equals(1, this.foo);
      expect$.Expect.equals(0, (this.foo = dart.notNull(this.foo) - 1));
      expect$.Expect.equals(0, this.foo);
      expect$.Expect.equals(0, (() => {
        let i = 0, x = this.get(i);
        this.set(i, dart.notNull(x) + 1);
        return x;
      })());
      expect$.Expect.equals(1, this.get(0));
      expect$.Expect.equals(2, (() => {
        let i = 0;
        return this.set(i, dart.notNull(this.get(i)) + 1);
      })());
      expect$.Expect.equals(2, this.get(0));
      expect$.Expect.equals(2, (() => {
        let i = 0, x = this.get(i);
        this.set(i, dart.notNull(x) - 1);
        return x;
      })());
      expect$.Expect.equals(1, this.get(0));
      expect$.Expect.equals(0, (() => {
        let i = 0;
        return this.set(i, dart.notNull(this.get(i)) - 1);
      })());
      expect$.Expect.equals(0, this.get(0));
      let $0 = 42, $1 = 87, $2 = 117;
      expect$.Expect.equals(42, $0++);
      expect$.Expect.equals(43, $0);
      expect$.Expect.equals(44, ++$0);
      expect$.Expect.equals(88, ($0 = $0 + $0));
      expect$.Expect.equals(87, $1++);
      expect$.Expect.equals(88, $1);
      expect$.Expect.equals(89, ++$1);
      expect$.Expect.equals(178, ($1 = $1 + $1));
      expect$.Expect.equals(117, $2++);
      expect$.Expect.equals(118, $2);
      expect$.Expect.equals(119, ++$2);
    }
    testMangling() {
      let $0 = 42, $1 = 87, $2 = 117;
      this.set(0, 0);
      expect$.Expect.equals(42, (() => {
        let i = 0;
        return this.set(i, dart.notNull(this.get(i)) + $0);
      })());
      expect$.Expect.equals(129, (() => {
        let i = 0;
        return this.set(i, dart.notNull(this.get(i)) + $1);
      })());
      expect$.Expect.equals(246, (() => {
        let i = 0;
        return this.set(i, dart.notNull(this.get(i)) + $2);
      })());
    }
  };
  dart.setSignature(expression_test.ExpressionTest, {
    constructors: () => ({new: dart.definiteFunctionType(expression_test.ExpressionTest, [])}),
    methods: () => ({
      testBinary: dart.definiteFunctionType(dart.dynamic, []),
      testUnary: dart.definiteFunctionType(dart.dynamic, []),
      testShifts: dart.definiteFunctionType(dart.dynamic, []),
      testBitwise: dart.definiteFunctionType(dart.dynamic, []),
      get: dart.definiteFunctionType(dart.dynamic, [core.int]),
      set: dart.definiteFunctionType(dart.dynamic, [core.int, core.int]),
      testIncrement: dart.definiteFunctionType(dart.dynamic, []),
      testMangling: dart.definiteFunctionType(dart.void, [])
    }),
    statics: () => ({testMain: dart.definiteFunctionType(dart.dynamic, [])}),
    names: ['testMain']
  });
  expression_test.main = function() {
    expression_test.ExpressionTest.testMain();
  };
  dart.fn(expression_test.main, VoidTodynamic());
  // Exports:
  exports.expression_test = expression_test;
});

File: tests/corelib_2/expression_test.dart
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import "package:expect/expect.dart";

// Tests basic expressions. Does not attempt to validate the details of arithmetic, coercion, and
// so forth.
class ExpressionTest {
  ExpressionTest() {}

  int foo;

  static testMain() {
    var test = new ExpressionTest();
    test.testBinary();
    test.testUnary();
    test.testShifts();
    test.testBitwise();
    test.testIncrement();
    test.testMangling();
  }

  testBinary() {
    int x = 4, y = 2;
    Expect.equals(6, x + y);
    Expect.equals(2, x - y);
    Expect.equals(8, x * y);
    Expect.equals(2, x / y);
    Expect.equals(0, x % y);
  }

  testUnary() {
    int x = 4, y = 2, z = -5;
    bool t = true, f = false;
    Expect.equals(-4, -x);
    Expect.equals(4, ~z);
    Expect.equals(f, !t);
  }

  testShifts() {
    int x = 4, y = 2;
    Expect.equals(y, x >> 1);
    Expect.equals(x, y << 1);
  }

  testBitwise() {
    int x = 4, y = 2;
    Expect.equals(6, (x | y));
    Expect.equals(0, (x & y));
    Expect.equals(6, (x ^ y));
  }

  operator [](int index) {
    return foo;
  }

  operator []=(int index, int value) {
    foo = value;
  }

  testIncrement() {
    int x = 4, a = x++;
    Expect.equals(4, a);
    Expect.equals(5, x);
    Expect.equals(6, ++x);
    Expect.equals(6, x++);
    Expect.equals(7, x);
    Expect.equals(6, --x);
    Expect.equals(6, x--);
    Expect.equals(5, x);

    this.foo = 0;
    Expect.equals(0, this.foo++);
    Expect.equals(1, this.foo);
    Expect.equals(2, ++this.foo);
    Expect.equals(2, this.foo);
    Expect.equals(2, this.foo--);
    Expect.equals(1, this.foo);
    Expect.equals(0, --this.foo);
    Expect.equals(0, this.foo);

    Expect.equals(0, this[0]++);
    Expect.equals(1, this[0]);
    Expect.equals(2, ++this[0]);
    Expect.equals(2, this[0]);
    Expect.equals(2, this[0]--);
    Expect.equals(1, this[0]);
    Expect.equals(0, --this[0]);
    Expect.equals(0, this[0]);

    int $0 = 42, $1 = 87, $2 = 117;
    Expect.equals(42, $0++);
    Expect.equals(43, $0);
    Expect.equals(44, ++$0);
    Expect.equals(88, $0 += $0);
    Expect.equals(87, $1++);
    Expect.equals(88, $1);
    Expect.equals(89, ++$1);
    Expect.equals(178, ($1 += $1));
    Expect.equals(117, $2++);
    Expect.equals(118, $2);
    Expect.equals(119, ++$2);
  }

  void testMangling() {
    int $0 = 42, $1 = 87, $2 = 117;
    this[0] = 0;
    Expect.equals(42, (this[0] += $0));
    Expect.equals(129, (this[0] += $1));
    Expect.equals(246, (this[0] += $2));
  }
}

main() {
  ExpressionTest.testMain();
}
